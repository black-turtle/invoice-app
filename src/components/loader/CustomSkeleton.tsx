import { Skeleton, Typography } from '@mui/material'
import Box from '@mui/material/Box'
import { ReactNode } from 'react'

interface IProps {
    cnt?: number
    isTable?: boolean
}

function getSkeletons(cnt: number, isTable?: boolean) {
    const nodes: ReactNode[] = []
    for (let i = 1; i <= cnt; i++) {
        nodes.push(
            <Typography key={i} variant={isTable ? 'h4' : 'h2'}>
                <Skeleton animation="wave" />
            </Typography>,
        )
    }
    return nodes
}
export const LOADING_SKELETON_TEST_ID = 'loading-skelton'
export default function CustomSkeleton({ cnt, isTable }: IProps) {
    return (
        <Box
            data-test={LOADING_SKELETON_TEST_ID}
            data-testid={LOADING_SKELETON_TEST_ID}
            sx={{ width: '100%', ...(!isTable && { my: '3rem' }) }}
        >
            {getSkeletons(cnt ?? 5, isTable)}
        </Box>
    )
}
