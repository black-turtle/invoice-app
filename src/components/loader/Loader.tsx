import { CircularProgress } from '@mui/material'

interface ProgressProps {
    // containerStyles?: Record<string, string>
    // progressProps?: Record<string, string>
    fullScreen?: boolean
}

const fullPageStyles = {
    container: {
        height: '100vh',
    },
    progressProps: {
        size: '10vmin',
    },
}

const Loader = ({ fullScreen: isFullPage }: ProgressProps) => {
    return (
        <div
            className="flex-center-container"
            style={isFullPage ? fullPageStyles.container : undefined}
        >
            <CircularProgress
                {...(isFullPage && fullPageStyles.progressProps)}
            />
        </div>
    )
}

export default Loader
