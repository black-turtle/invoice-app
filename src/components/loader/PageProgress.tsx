import NProgress from 'nprogress'
import 'nprogress/nprogress.css'
import { useEffect } from 'react'
import useNavigate from '../../hooks/useNavigate'

NProgress.configure({ showSpinner: false, easing: '' })

const PageChangeLoader = () => {
    const { router } = useNavigate()

    // show progress bar while page changing
    useEffect(() => {
        const handleStart = (url: string) => {
            NProgress.start()
        }

        const handleStop = () => {
            NProgress.done()
        }

        router.events.on('routeChangeStart', handleStart)
        router.events.on('routeChangeComplete', handleStop)
        router.events.on('routeChangeError', handleStop)

        return () => {
            router.events.off('routeChangeStart', handleStart)
            router.events.off('routeChangeComplete', handleStop)
            router.events.off('routeChangeError', handleStop)
        }
    }, [router])

    // this component doesn't display anything
    return null
}

export default PageChangeLoader
