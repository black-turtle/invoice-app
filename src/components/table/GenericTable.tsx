import { Box } from '@mui/material'
import Paper from '@mui/material/Paper'
import Table from '@mui/material/Table'
import TableContainer from '@mui/material/TableContainer'
import { ReactNode } from 'react'

interface GenericTableProps {
    dataTestAttr?: string
    tableFilter?: ReactNode
    tableHeader?: ReactNode
    tableBody?: ReactNode
    tablePagination?: ReactNode
}

export default function GenericTable(props: GenericTableProps) {
    return (
        <Box sx={{ mb: '5rem' }}>
            {props.tableFilter && props.tableFilter}
            {props.tablePagination && props.tablePagination}

            <TableContainer
                component={Paper}
                data-test={props.dataTestAttr}
                data-testid={props.dataTestAttr}
            >
                <Table aria-label="table">
                    {props.tableHeader && props.tableHeader}

                    {props.tableBody && props.tableBody}
                </Table>
            </TableContainer>
        </Box>
    )
}
