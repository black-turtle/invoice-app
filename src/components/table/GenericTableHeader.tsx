import {
    styled,
    TableCell,
    tableCellClasses,
    TableHead,
    TableRow,
    TableSortLabel,
    tableSortLabelClasses,
    Typography,
} from '@mui/material'

const StyledTableCell = styled(TableCell)(({ theme }) => ({
    [`&.${tableCellClasses.head}`]: {
        backgroundColor: theme.palette.primary.main,
        color: theme.palette.common.white,
        textTransform: 'uppercase',
        fontWeight: 'bold',
        fontSize: '1.2rem',
    },
}))

const StyledTableSortLabel = styled(TableSortLabel)(({ theme }) => ({
    [`& .${tableSortLabelClasses.icon}`]: {
        color: `${theme.palette.common.white} !important`,
    },
}))

export interface TableHeaderLabelData {
    label: string
    isSortable?: boolean
    dataTestAttr?: string
}

export interface TableHeaderData {
    headers: Record<string, TableHeaderLabelData>
    sortEnabled?: boolean
    sort?: 'asc' | 'desc'
    sortBy?: string
    rowActionRequired?: boolean
    onSortChangeAction?: (
        updateQuery: Record<string, string | undefined>,
    ) => void
}

const GenericTableHeader = (props: TableHeaderData) => {
    const changeSortDirection = (newSortBy: string) => {
        let { sort, sortBy } = props
        if (newSortBy === sortBy) {
            sort = sort === 'asc' ? 'desc' : 'asc'
        }
        props.onSortChangeAction!({
            sortBy: newSortBy,
            sort,
            offset: undefined,
        })
    }

    const { headers, rowActionRequired, sort, sortBy, sortEnabled } = props
    return (
        <TableHead>
            <TableRow>
                {Object.keys(headers).map((key, idx) => (
                    <StyledTableCell
                        key={`header-${key}`}
                        align={idx > 0 ? 'right' : 'left'}
                        data-test={headers[key].dataTestAttr}
                        data-testid={headers[key].dataTestAttr}
                    >
                        {/* Apply sort label only if sortEnabled and row is sortable  */}
                        {sortEnabled && headers[key].isSortable && (
                            <StyledTableSortLabel
                                sx={{ color: 'azure !important' }}
                                active={key === sortBy}
                                direction={key === sortBy ? sort : undefined}
                                onClick={() => changeSortDirection(key)}
                            >
                                {headers[key].label}
                            </StyledTableSortLabel>
                        )}

                        {!(sortEnabled && headers[key].isSortable) && (
                            <Typography sx={{ cursor: 'not-allowed' }}>
                                {headers[key].label}
                            </Typography>
                        )}
                    </StyledTableCell>
                ))}
                {rowActionRequired && <StyledTableCell />}
            </TableRow>
        </TableHead>
    )
}

export default GenericTableHeader
