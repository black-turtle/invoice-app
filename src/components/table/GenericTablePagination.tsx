import { TablePagination } from '@mui/material'

export interface TablePaginationData {
    offset: number
    limit: number
    total: number
    onChange?: (updateQuery: Record<string, string | undefined>) => void
}

const GenericTablePagination = ({
    offset,
    limit,
    total,
    onChange,
}: TablePaginationData) => {
    const handlePageChange = (newPage: number) => {
        onChange!({
            offset: (newPage * limit).toString(),
        })
    }

    const handleRowsPerPageChange = (newRowsPerPage: number) => {
        onChange!({
            limit: newRowsPerPage.toString(),
            offset: undefined,
        })
    }

    return (
        <TablePagination
            component="div"
            data-test="table-pagination"
            count={total}
            rowsPerPage={limit}
            page={offset / limit}
            onPageChange={(event, newPage) => {
                handlePageChange(newPage)
            }}
            onRowsPerPageChange={(event) => {
                handleRowsPerPageChange(Number(event.target.value))
            }}
        />
    )
}

export default GenericTablePagination
