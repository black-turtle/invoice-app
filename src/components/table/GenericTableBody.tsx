import ViewListIcon from '@mui/icons-material/ViewList'
import { TableBody } from '@mui/material'
import { styled } from '@mui/material/styles'
import TableCell, { tableCellClasses } from '@mui/material/TableCell'
import TableRow from '@mui/material/TableRow'

import CustomMenu, { MenuActionData } from '../CustomMenu'
import CustomSkeleton from '../loader/CustomSkeleton'

const StyledTableCell = styled(TableCell)(({ theme }) => ({
    [`&.${tableCellClasses.body}`]: {
        fontSize: 14,
        // padding: '0px 5px 0px 5px',
    },
}))

const StyledTableRow = styled(TableRow)(({ theme }) => ({
    '&:nth-of-type(odd)': {
        backgroundColor: theme.palette.action.hover,
    },
    // hide last border
    '&:last-child td, &:last-child th': {
        border: 0,
    },
}))
const rowHoverStyle = {
    '&:hover': {
        backgroundColor: (theme: any) => theme.palette.action.selected,
        cursor: 'pointer',
    },
}

export interface TableRowData {
    rowKey: string | number
    rowUrl?: string
    // data: Record<string, string | number | undefined>
    data: Record<
        string,
        {
            value: string | number | undefined
            dataTestAttr?: string
        }
    >
    actions?: {
        dataTestAttr?: string
        data: MenuActionData[]
    }
}

const EmptyTableBody = ({
    colSpan,
    rows,
}: {
    rows: TableRowData[] | null
    colSpan: number
}) => {
    return (
        <TableBody>
            <StyledTableRow>
                <StyledTableCell rowSpan={5} colSpan={colSpan} align="center">
                    {!rows && <CustomSkeleton cnt={3} isTable />}
                    {rows?.length === 0 && (
                        <img
                            src="/images/no-data.png"
                            alt="empty-table"
                            style={{ height: '30vh' }}
                            data-test="empty-placeholder"
                            data-testid="empty-placeholder"
                        />
                    )}
                </StyledTableCell>
            </StyledTableRow>
        </TableBody>
    )
}

const GenericTableBody = ({
    rows,
    headerKeys,
    rowClickAction,
}: {
    rows: TableRowData[] | null
    headerKeys: string[]
    rowClickAction?: (newPath?: string | undefined) => void
}) => {
    if (!rows || rows.length === 0) {
        return <EmptyTableBody rows={rows} colSpan={headerKeys.length + 1} />
    }

    return (
        <TableBody>
            {rows.map((row, rowIdx) => (
                <StyledTableRow
                    key={row.rowKey}
                    data-test={row.rowKey}
                    sx={{ ...(row.rowUrl && rowHoverStyle) }}
                    onClick={() => rowClickAction && rowClickAction(row.rowUrl)}
                >
                    {headerKeys.map((key, colIdx) => (
                        <StyledTableCell
                            key={key}
                            component="th"
                            scope="row"
                            align={colIdx > 0 ? 'right' : 'left'}
                            data-test={row.data[key]?.dataTestAttr}
                        >
                            {row.data[key]?.value}
                        </StyledTableCell>
                    ))}

                    {row.actions && (
                        <StyledTableCell align="right">
                            <CustomMenu
                                header={<ViewListIcon />}
                                actions={row.actions.data}
                                dataTestAttr={row.actions.dataTestAttr}
                            />
                        </StyledTableCell>
                    )}
                </StyledTableRow>
            ))}
        </TableBody>
    )
}

export default GenericTableBody
