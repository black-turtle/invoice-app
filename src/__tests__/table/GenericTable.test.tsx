import '@testing-library/jest-dom'
import { render, screen } from '@testing-library/react'
import { LOADING_SKELETON_TEST_ID } from '../../components/loader/CustomSkeleton'

import GenericTable from '../../components/table/GenericTable'
import GenericTableBody, {
    TableRowData,
} from '../../components/table/GenericTableBody'
import GenericTableHeader from '../../components/table/GenericTableHeader'
import GenericTablePagination from '../../components/table/GenericTablePagination'

// Constants
const dataTestAttr = 'test-table'

const headerData = {
    header_1: {
        label: 'header_1',
        dataTestAttr: 'header_1',
    },
    header_2: {
        label: 'header_2',
        dataTestAttr: 'header_2',
    },
}

// table body data
const commonTableBodyData = {
    headerKeys: ['header_1', 'header_2'],
    rowClickAction: (url: string | undefined) =>
        console.log('navigate to ' + url),
}
const loadingTableBodyData = {
    rows: null,
    ...commonTableBodyData,
}

const emptyTableBodyData = {
    rows: [],
    ...commonTableBodyData,
}

const tableBodyData = {
    rows: [
        {
            rowKey: 'row-key',
            rowUrl: 'row-url',
            data: {
                header_1: {
                    value: 'data_1',
                    dataTestAttr: 'data_1_test',
                },
                header_2: {
                    value: 'data_2',
                    dataTestAttr: 'data_2_test',
                },
            },
        },
    ] as TableRowData[],
    ...commonTableBodyData,
}

describe('Test GenericTable', () => {
    test('Check table has exact data-test attribute', () => {
        render(<GenericTable dataTestAttr={dataTestAttr} />)
        expect(screen.getByTestId(dataTestAttr)).toBeInTheDocument()
    })

    test('test table loading state when data=null', () => {
        render(
            <GenericTable
                dataTestAttr={dataTestAttr}
                tableBody={<GenericTableBody {...loadingTableBodyData} />}
            />,
        )
        expect(screen.getByTestId(LOADING_SKELETON_TEST_ID)).toBeInTheDocument()
    })

    test('Check empty placeholder present if data = []', () => {
        render(
            <GenericTable
                dataTestAttr={dataTestAttr}
                tableBody={<GenericTableBody {...emptyTableBodyData} />}
            />,
        )
        expect(screen.getByTestId('empty-placeholder')).toBeInTheDocument()
    })

    test('Check all table data displayed properly if row data present', () => {
        render(
            <GenericTable
                dataTestAttr={dataTestAttr}
                tableBody={<GenericTableBody {...tableBodyData} />}
            />,
        )

        // check every data key present
        tableBodyData.rows.forEach((row) => {
            Object.keys(row.data).forEach((colData) => {
                expect(
                    screen.getByText(row.data[colData].value as string),
                ).toBeInTheDocument()
            })
        })
    })

    test('test tableHeader displaying expected result', () => {
        render(
            <GenericTable
                dataTestAttr={dataTestAttr}
                tableHeader={<GenericTableHeader headers={headerData} />}
            />,
        )

        // check header label displayed
        expect(screen.getByText('header_2')).toBeInTheDocument()

        // check header testId present
        expect(screen.getByTestId('header_2')).toBeInTheDocument()
    })

    test('test table Pagination displaying expected results', () => {
        render(
            <GenericTable
                dataTestAttr={dataTestAttr}
                tablePagination={
                    <GenericTablePagination offset={0} limit={10} total={100} />
                }
            />,
        )

        expect(screen.getByText('1â€“10 of 100')).toBeInTheDocument()
    })

    test('Full Snapshot test', async () => {
        const { container } = render(
            <GenericTable dataTestAttr={dataTestAttr} />,
        )

        expect(container).toMatchSnapshot()
    })
})
