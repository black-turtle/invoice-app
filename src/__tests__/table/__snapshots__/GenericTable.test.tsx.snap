// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test GenericTable Full Snapshot test 1`] = `
<div>
  <div
    class="MuiBox-root css-nd92od"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-11xur9t-MuiPaper-root-MuiTableContainer-root"
      data-test="test-table"
      data-testid="test-table"
    >
      <table
        aria-label="table"
        class="MuiTable-root css-rqglhn-MuiTable-root"
      />
    </div>
  </div>
</div>
`;
